test_communication/test/test_action_client.cpp

```diff
@@ -0,0 +1,193 @@
+  const std::string & action_type_name,
+  const std::vector<ActionClientTest<ActionT>> & goal_tests)
+{
+  auto action_client =
+    rclcpp_action::create_client<ActionT>(node, "test/action/" + action_type_name);
+  auto logger = node->get_logger();
+
+  if (!action_client->wait_for_action_server(20s)) {
+    RCLCPP_ERROR(logger, "requester service not available after waiting");
+    return 1;
+  }
+
+  size_t test_index = 0;
+  bool invalid_feedback = false;
+  auto start = std::chrono::steady_clock::now();
+  RCLCPP_SCOPE_EXIT({
+    auto end = std::chrono::steady_clock::now();
+    std::chrono::duration<float> diff = (end - start);
+    RCLCPP_INFO(logger, "sent goals for %f seconds\n", diff.count());
+  });
+
+  while (rclcpp::ok() && test_index < goal_tests.size()) {
+    RCLCPP_INFO(logger, "sending goal #%zu", test_index + 1);
+
+    // on feedback, check the feedback is valid
+    auto feedback_callback =
+      [&](auto, const auto & feedback) {
+        RCLCPP_INFO(logger, "received feedback");
+        if (!goal_tests[test_index].feedback_is_valid(feedback)) {
+          RCLCPP_ERROR(logger, "invalid feedback");
+          invalid_feedback = true;
+        }
+      };
+
+    // send the request
+    auto goal_handle_future =
+      action_client->async_send_goal(goal_tests[test_index].goal, feedback_callback);
+
+    using rclcpp::executor::FutureReturnCode;
+    // wait for the sent goal to be accepted
+    auto status = rclcpp::spin_until_future_complete(node, goal_handle_future, 1000s);
+    if (status != FutureReturnCode::SUCCESS) {
+      RCLCPP_ERROR(logger, "send goal call failed");
+      return 1;
+    }
+
+    // wait for the result (feedback may be received in the meantime)
+    auto result_future = goal_handle_future.get()->async_result();
+    status = rclcpp::spin_until_future_complete(node, result_future, 1000s);
+    if (status != FutureReturnCode::SUCCESS) {
+      RCLCPP_ERROR(logger, "failed to receive a goal result in time");
+      return 1;
+    }
+
+    if (!goal_tests[test_index].result_is_valid(result_future.get().response)) {
+      RCLCPP_ERROR(logger, "invalid goal result");
+      return 1;
+    }
+    RCLCPP_INFO(logger, "received goal #%zu of %zu", test_index + 1, goal_tests.size());
+    test_index++;
+  }
+
+  if (test_index != goal_tests.size()) {
+    return 1;
+  }
+
+  if (invalid_feedback) {
+    return 1;
+  }
+
+  return 0;
+}
+
+std::vector<ActionClientTest<test_msgs::action::Fibonacci>>
+generate_fibonacci_goal_tests()
+{
+  std::vector<ActionClientTest<test_msgs::action::Fibonacci>> result;
+
+  constexpr size_t order = 10;
+  static_assert(order > 0, "order needs to be non-zero");
+
+  std::vector<int32_t> valid_fibo_seq;
+  valid_fibo_seq.push_back(0);
+  valid_fibo_seq.push_back(1);
+  for (size_t i = 1; i < order; ++i) {
+    valid_fibo_seq.push_back(valid_fibo_seq[i] + valid_fibo_seq[i - 1]);
+  }
+
+  {
+    ActionClientTest<test_msgs::action::Fibonacci> test;
+    size_t order = 10;
```

Comment:
@wjwwood hiding above's variable?

Reply:
Yeah, that was intentional, the idea being each scoped section would make their own order, but I guess I could use different names here.